polynote-frontend/polynote/messages.js

```diff
@@ -211,6 +211,34 @@ export class CellResult extends Message {
 CellResult.codec = combined(shortStr, int16, Result.codec).to(CellResult);
 
 
+export class NotebookUpdate extends Message {
+
+    static unapply(inst) { return [inst]; }
+
+    /**
+     * Transform a so that it has the same effect when applied after b. Returns transformed a.
+     * @seeScala polynote.messages.NotebookUpdate#rebase
+     */
+    static rebase(a, b) {
+        if (b instanceof Array) {
+            let accum = a;
+            b.forEach(update => {
+                accum = NotebookUpdate.rebase(accum, update);
+            });
+            return accum;
+        }
+
+        if (a instanceof InsertCell && b instanceof InsertCell && a.after === b.after) {
+            return new InsertCell(a.path, a.globalVersion, a.localVersion, b.cell.id);
+        } else if (a instanceof UpdateCell && b instanceof UpdateCell && a.id === b.id) {
+            return new UpdateCell(a.path, a.globalVersion, a.localVersion, ContentEdit.rebaseEdits(a.edits, b.edits));
+        } else {
+            return a;
```

Comment:
So if `b` is neither an `Array` nor an `InsertCell` nor an `UpdateCell` we just ignore it and return `a` - is there any case we expect this to be the case under normal operation? If not, maybe we should log something?

Reply:
It's expected to happen all the time - it just handles any case where `a` and `b` don't conflict (or we don't have a way of handling the conflict anyway). There's a comment to that effect in the corresponding Scala code.

Reply:
🤦‍♂️